File: alymov/lab_5/solver.py
import numpy as np

def Tridiagonal_matrix_algorithm(L, y):
    a = [0]
    b =[L[0][0]]
    c = [L[0][1]]
    n = len(y)

    x = []
    alpha = []
    beta = []

    for i in range(1, n - 1):
        a.append(L[i][i-1])
        b.append(L[i][i])
        c.append(L[i][i+1])

    a.append(L[n-1][n-2])
    b.append(L[n-1][n-1])
    c.append(0)

    a = np.array(a)
    b = np.array(b)
    c = np.array(c)

    alpha.append(-c[0]/b[0])
    beta.append(y[0]/b[0])

    for i in range(1, n-1):
        d = a[i] * alpha[i-1] + b[i]
        alpha.append(-c[i] / d)
        beta.append((y[i] - a[i] * beta[i-1]) / d)
    alpha.append(0)
    beta.append((y[n - 1] - a[n-1] * beta[n-2]) / (a[n-1] * alpha[n-2] + b[n-1]))

    x.append(beta[n-1])

    for i in range(1, n):
        x.append(x[i - 1] * alpha[n - 1 - i] + beta[n - 1 - i])

    return x[::-1]

File: lab_6/solver.cpp
#include "matrix.hpp"
 
Matrix Tridiagonal_matrix_algorithm(Matrix A, Vector b)
{

    int n = A.rows;
    Vector alpha(n);
    Vector beta(n);
    Vector y(n);
    Vector x(n);
    
    // A[i] = A[i][i-1]
    // B[i] = A[i][i+1]
    // C[i] = A[i][i+1]
    y[0] = A[0][0];
    alpha[0] = -A[0][1] / y[0];
    beta[0] = b[0] / y[0];

    for(int i = 1; i < n-1; i++)
    {
        y[i] = alpha[i - 1] * A[i][i - 1] + A[i][i];
        alpha[i] = -A[i][i + 1] / y[i];
        beta[i] = (b[i] - A[i][i-1] * beta[i - 1]) / y[i];
    }
 
    y[n - 1] = alpha[n - 2] * A[n - 1][n - 2] + A[n - 1][n - 1];
    beta[n - 1] = (b[n - 1] - A[n-1][n-2] * beta[n - 2]) / y[n - 1];

    x[n-1] = beta[n-1];

    for(int i = n - 2; i >= 0; i--)
    {
        x[i] = alpha[i] * x[i + 1] + beta[i];
    }

    return x;
}